src/Language/DifferentialDatalog/Crate.hs

```diff
@@ -0,0 +1,235 @@
+crateName crate | elem (crateMainModule crate) stdLibs =
+    intercalate "__" $ (modulePath $ crateMainModule crate)
+                | otherwise =
+    intercalate "__" $ "types" : (modulePath $ crateMainModule crate)
+
+-- A crate graph is a graph with crates as vertices and with two
+-- types of edges:
+-- * Dependency edges represent module dependencies.  Two crates are connected
+--   by a dependency edge iff there is a dependency between a pair of modules
+--   inside these crates.
+-- * Parent edges represent the DDlog module hierarchy.
+data CrateGraph = CrateGraph {
+    -- Crates.
+    cgCrates    :: [Crate],
+    -- Map from module names to index of the crate the module belongs to.
+    cgMod2Crate :: M.Map ModuleName Int
+} deriving (Eq, Show)
+
+cgEmpty :: CrateGraph
+cgEmpty =
+    CrateGraph {
+        cgCrates       = [],
+        cgMod2Crate    = M.empty
+    }
+
+cgAddCrate :: CrateGraph -> Crate -> CrateGraph
+cgAddCrate cg@CrateGraph{..} crate | elem crate cgCrates = cg
+                                   | otherwise =
+    CrateGraph {
+        cgCrates = cgCrates ++ [crate],
+        cgMod2Crate = foldl' (\modCrates m -> M.insert m (length cgCrates) modCrates)
+                             cgMod2Crate (S.toList crate)
+    }
+
+cgLookupCrate :: CrateGraph -> String -> Maybe Crate
+cgLookupCrate cg crate_name = find ((== crate_name) . crateName) $ cgCrates cg
+
+cgModuleCrate :: CrateGraph -> ModuleName -> Crate
+cgModuleCrate CrateGraph{..} mod_name = cgCrates !! (cgMod2Crate M.! mod_name)
+
+-- Module dependencies.  Currently just the list of module imports, but
+-- we may want to use a more accurate method based on actual dependencies
+-- used by types and functions in the module.
+moduleDeps :: (?mmap :: M.Map ModuleName DatalogModule) => ModuleName -> [ModuleName]
+moduleDeps m = fromMaybe [] $ ((map importModule) . progImports . moduleDefs) <$> ?mmap M.!? m
+
+-- Expand crate so that modules in the crate form a single connected
+-- subtree of the module graph:
+-- Find the lowest common ancestor of all modules in the crate and add
+-- modules along all branches from the ancestor to modules in the crate to
+-- the crate.
+--
+-- Example:
+-- Module hirarchy:
+-- top
+--  |
+--  |-m1
+--    |-m2
+--    |-m3
+--    |  |-m4
+--    |  |-m5
+--    |
+--    |-m6
+--       |-m7
+--       |-m8
+--
+-- crate1 = {m8, m4}.
+-- The lowest common ancestor of m8 and m4 is m1.  We therefore
+-- extend the crate with m1 and all modules down the path from m1
+-- to m8 and m4:
+-- crate1' = {m1, m3, m4, m6, m8}.
+addLowestCommonAncestor :: Crate -> Crate
+addLowestCommonAncestor crate =
+    S.foldl (\mods m -> foldl' (\mods' mname -> S.insert (ModuleName mname) mods') (S.insert m mods)
+                               $ drop lca_len $ inits $ modulePath m)
+            S.empty crate
+    where
+    -- Lowest common ancestor path length.
+    lca_len = length $ lcp $ map modulePath $ S.toList crate
+    -- Longest common prefix.
```

Comment:
these could be utility functions in some library, but I don't know if it's worth moving them